specs/Standards/Tokens/FungibleTokenMetadata.md

```diff
@@ -0,0 +1,99 @@
+
+A contract may update any metadata fields with the exception of `decimals`. A contract MUST NOT update `decimals`, as this can effectively change the balance of all holders.
+
+**Interface**:
+
+```ts
+type FungibleTokenMetadata = {
+    spec: string;
+    name: string;
+    symbol: string;
+    icon: string|null;
+    reference: string|null;
+    reference_hash: string|null;
+    decimals: number;
+}
+```
+
+**An implementing contract MUST include the following fields on-chain**
+
+- `spec`: a string. Should be `ft-1.0.0` to indicate that a Fungible Token contract adheres to the current versions of this Metadata and the [Fungible Token Core](./FungibleTokenCore.md) specs. This will allow consumers of the Fungible Token to know if they support the features of a given contract.
+- `name`: the human-readable name of the token.
+- `symbol`: the abbreviation, like wETH or AMPL.
+- `decimals`: used in frontends to show the proper significant digits of a token. This concept is explained well in this [OpenZeppelin post](https://docs.openzeppelin.com/contracts/3.x/erc20#a-note-on-decimals).
+
+**An implementing contract MAY include the following fields on-chain**
+
+- `icon`: a small image associated with this token. Must be a [data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs), to help consumers display it quickly while protecting user data. Recommendation: use [optimized SVG](https://codepen.io/tigt/post/optimizing-svgs-in-data-uris), which can result in high-resolution images with only 100s of bytes of [storage cost](https://docs.near.org/docs/concepts/storage-staking). (Note that these storage costs are incurred to the token owner/deployer, but that querying these icons is a very cheap & cachable read operation for all consumers of the contract and the RPC nodes that serve the data.) Recommendation: create icons that will work well with both light-mode and dark-mode websites by either using middle-tone color schemes, or by [embedding `media` queries in the SVG](https://timkadlec.com/2013/04/media-queries-within-svg/).
+- `reference`: a link to a valid JSON file containing various keys offering supplementary details on the token. Example: "/ipfs/QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm", "https://example.com/token.json", etc. If the information given in this document conflicts with the on-chain attributes, the values in `reference` shall be considered the source of truth.
+- `reference_hash`: the base64-encoded sha256 hash of the JSON file contained in the `reference` field. This is to guard against off-chain tampering.
+
+## Drawbacks
+
+- Enforcement of `icon` to be a data URL rather than a link to an HTTP endpoint that could contain privacy-violating code cannot be done on deploy or update of contract metadata, and must be done on the consumer/app side when displaying token data.
```

Comment:
I disagree with the icon has to be a data URL. You need to be experience web designed to craft a SVG icon. Safari still doesn't support `webp`, so your safest fall back is data URI encoded `png` or `gif` which is still a few KB of data on chain. 
Brave enforces `https` on the image urls, so you can't really mess the privacy. If you're fetching the reference, it's also violates privacy. The app may decide to use image proxy if needed.

Reply:
Agreed, and we're doing it the way Eugene describes.

Reply:
> If you're fetching the reference, it's also violates privacy. 

Yes, this is true, and something we should make application developers aware of.

> You need to be experience web designed to craft a SVG icon

I think we should set a high-but-reasonable bar in the recommendation. This was a concern that was brought up in the Feb 26 call. 

Contract development teams are free to turn PNG or GIF images into data URLs instead, if they can't afford to contract a designer for one asset and can't figure out how to use ImageMagick to turn their PNG into an SVG.

Reply:
Might need to update where the metadata is stored in the contract in the token example, so we lazy load it in the contract. We may want to add a new collection. Like `Lazy<T: BorshSerialize+BorshDeserialize>`

Reply:
> I disagree with the icon has to be a data URL. You need to be experience web designed to craft a SVG icon. Safari still doesn't support `webp`, so your safest fall back is data URI encoded `png` or `gif` which is still a few KB of data on chain.
> Brave enforces `https` on the image urls, so you can't really mess the privacy. If you're fetching the reference, it's also violates privacy. The app may decide to use image proxy if needed.

Not sure how https helps here at all. The problem is privacy from the creators of tokens. Creators of token can track everybody who loads the icon if it’s web URL, no matter whether HTTP or HTTPS. 

like you can of course decide not to change standard. But in that case we not gonna use icon field in wallet at all.

Reply:
> The app may decide to use image proxy if needed.

this doesn’t really help with airdropped tokens, as you can airdrop with custom contract and so icon URL for every user

Reply:
I'm now understanding what you're talking about. I think I needed an example for it to sink in. Some hacker could write a fungible token contract that has an icon that points to https://example.com/mike-purvis-opened-this.jpg and send me an email with a link to https://wallet.example.com/accounts/fake-token-182537893 and when I go there, the Wallet site requests mike-purvis-opened-this.jpg and the web server remembers this and then knows that I followed the link via email